File: utils/src/main/java/io/github/aemogie/timble/utils/events/EventBus.java
package io.github.aemogie.timble.utils.events;

import io.github.aemogie.timble.utils.events.Event.Listener;

import java.util.ArrayList;
import java.util.HashMap;

public class EventBus {
	private EventBus() {}
	
	static final HashMap<Class<Event>, ArrayList<Listener<Event>>> LISTENERS = new HashMap<>();
	
	public static <T extends Event> boolean fireEvent(T event) {
		return event.initFire() && event.fire(LISTENERS.get(event.getClass()));
	}
	
	@SuppressWarnings("unchecked")
	public static <T extends Event> Listener<T> subscribeToEvent(Class<T> clazz, Listener<T> listener) {
		LISTENERS.computeIfAbsent((Class<Event>) clazz, eventClass -> new ArrayList<>());
		LISTENERS.get(clazz).add((Listener<Event>) listener);
		return listener;
	}
}

File: utils/src/main/kotlin/io/github/aemogie/timble/utils/events/EventBus.kt
//EventBus = singleton of EventNode
@file:JvmName("EventBus")
package io.github.aemogie.timble.utils.events

import kotlin.reflect.KClass
private val EVENT_BUS = EventNode()
fun <T : Event> fireEvent(event: T) = EVENT_BUS.fire(event)
fun <T : Event> subscribeToEvent(clazz: Class<T>, listener: (T) -> Boolean) = EVENT_BUS.subscribe(clazz, listener)
fun <T : Event> subscribeToEvent(clazz: KClass<T>, listener: (T) -> Boolean) = EVENT_BUS.subscribe(clazz, listener)